packages/utilities/src/inferLanguages.ts

```diff
@@ -0,0 +1,122 @@
+import {
+    Concept,
+    Containment,
+    Language,
+    Link,
+    Property,
+    Reference,
+    SerializationChunk,
+    builtinPrimitives,
+    chain,
+    concatenator,
+    lastOf
+} from "@lionweb/core"
+import { hasher } from "./hashing.js"
+
+const id = chain(concatenator("-"), hasher())
+const key = lastOf
+
+const { stringDatatype } = builtinPrimitives
+
+export function inferLanguagesFromChunk(chunk: SerializationChunk): Language[] {
+    const languages = new Map<string, Language>()
+    const concepts = new Map<string, Concept>()
+    const links = new Array<{ link: Link; conceptId: string }>()
+
+    for (const chunkLanguage of chunk.languages) {
+        const languageName = chunkLanguage.key
+        const language = new Language(languageName, chunkLanguage.version, id(languageName), key(languageName))
+        languages.set(languageName, language)
+    }
+
+    for (const node of chunk.nodes) {
+        const languageName = node.classifier.language
+        const entityName = node.classifier.key
+
+        const language = getLanguage(languages, languageName)
+        if (language.entities.filter(entity => entity.key === entityName).length) {
+            continue
+        }
+
+        const concept = new Concept(language, entityName, key(language.name, entityName), id(language.name, entityName), false)
+        language.havingEntities(concept)
+        concepts.set(node.id, concept)
+
+        for (const property of node.properties) {
+            const propertyName = property.property.key.split("_")[2]
```

Comment:
Does this work in general? It assumes that keys have the form `<0>_<1>_<2>_...` and that the <2> part is the only meaningful part, right?

Reply:
That's how I interpret this key `library_Library_name`. The first part is the language key, the second part is the classifier key, and the third part is the property key. Do I miss something?

Reply:
Any language can choose how its keys itself. The pattern is true for some of our example languages (like library), but not even for the LionCore M3 and built-ins languages (which use "`-`" as the separator). So, in general, this doesn't work.

OTOH: it's the only thing you can do to come up with some kind of meaningful name. Maybe you could extend this a bit with trying to split on various kinds of separators ("`-`", "`_`"), and seeing whether that split matches some kind of notion of a qualified name? But that already suggests a lot of heuristics, so maybe we should stick to something simple/easy that works 70% of the time.